fdbserver/ApplyMetadataMutation.cpp

```diff
@@ -43,685 +43,828 @@ Reference<StorageInfo> getStorageInfo(UID id,
-						// Parse the backup key and name
-						logKeyBegin = logRangesDecodeKey(logRangeAffected.key, nullptr);
-
-						// Decode the log destination and key value
-						logKeyEnd = logRangesDecodeValue(logRangeAffected.value, &logDestination);
-
-						TraceEvent("LogRangeErase")
-						    .detail("AffectedKey", logRangeAffected.key)
-						    .detail("AffectedValue", logRangeAffected.value)
-						    .detail("LogKeyBegin", logKeyBegin)
-						    .detail("LogKeyEnd", logKeyEnd)
-						    .detail("LogDestination", logDestination);
-
-						// Identify the locations to place the backup key
-						auto logRanges = vecBackupKeys->modify(KeyRangeRef(logKeyBegin, logKeyEnd));
-
-						// Remove the log prefix from the ranges which include it
-						for (auto logRange : logRanges) {
-							auto& logRangeMap = logRange->value();
-
-							// Remove the backup name from the range
-							logRangeMap.erase(logDestination);
-						}
 
-						bool foundKey = false;
-						for (auto& it : vecBackupKeys->intersectingRanges(normalKeys)) {
-							if (it.value().count(logDestination) > 0) {
-								foundKey = true;
-								break;
-							}
-						}
-						if (!foundKey) {
-							auto logRanges = vecBackupKeys->modify(singleKeyRange(metadataVersionKey));
-							for (auto logRange : logRanges) {
-								auto& logRangeMap = logRange->value();
-								logRangeMap.erase(logDestination);
-							}
-						}
-					}
+			// Coallesce the entire range
+			vecBackupKeys->coalesce(allKeys);
+		}
 
-					// Coallesce the entire range
-					vecBackupKeys->coalesce(allKeys);
-				}
+		if (!initialCommit)
+			txnStateStore->clear(commonLogRange);
+	}
 
-				if (!initialCommit)
-					txnStateStore->clear(commonLogRange);
-			}
+	void checkClearTssMappingKeys(MutationRef m, KeyRangeRef range) {
+		if (!tssMappingKeys.intersects(range)) {
+			return;
+		}
+		KeyRangeRef rangeToClear = range & tssMappingKeys;
+		ASSERT(rangeToClear.singleKeyRange());
+
+		// Normally uses key backed map, so have to use same unpacking code here.
+		UID ssId = Codec<UID>::unpack(Tuple::unpack(m.param1.removePrefix(tssMappingKeys.begin)));
+		if (!initialCommit) {
+			txnStateStore->clear(rangeToClear);
+		}
+
+		if (tssMapping) {
+			tssMapping->erase(ssId);
+		}
+
+		if (!toCommit) {
+			return;
+		}
+		// send private mutation to SS to notify that it no longer has a tss pair
+		if (Optional<Value> tagV = txnStateStore->readValue(serverTagKeyFor(ssId)).get(); tagV.present()) {
+			MutationRef privatized = m;
+			privatized.param1 = m.param1.withPrefix(systemKeys.begin, arena);
+			toCommit->addTag(decodeServerTagValue(tagV.get()));
+			toCommit->writeTypedMessage(privatized);
 		}
 	}
 
-	for (KeyRangeRef& range : tssServerListToRemove) {
-		txnStateStore->clear(range);
+	void checkClearTssQuarantineKeys(MutationRef m, KeyRangeRef range) {
+		if (!tssQuarantineKeys.intersects(range)) {
+			return;
+		}
+		if (initialCommit) {
+			return;
```

Comment:
Nit: these two conditions can be combined.

Reply:
Updated.